man/gpasm.1
.TH GPASM 1 "2007-10-27" "gputils-2.0.0_beta2" "Linux user's manual"
.SH NAME
gpasm \- GNU PIC assembler
.SH SYNOPSIS
.B gpasm [options] file
.SH WARNING
The information in this man page is an extract from the full documentation of
gputils and is limited to the meaning of the options.  For complete and
current documentation, refer to "gputils.ps" located in the gputils docs
directory.
.SH DESCRIPTION
.B gpasm
is an assembler for Microchip (TM) PIC (TM) micro-controllers.
It's intended to be compatible with the manufacturer's MPASM
assembler.
.B gpasm
is part of gputils.  Check the
.BR gputils (1)
manpage for details on other GNU PIC utilities.
.SH OPTIONS
Below is a summary of the
.B gpasm
options.
.B gpasm
options always supersede any settings in the source code.
.TP
.BR "\-a FMT" , " \-\-hex\-format FMT
GPASM supports inhx8m, inhx8s, inhx16, and inhx32 hex file formats.  This
option controls which hex file format is used.  The default is inhx32.
.TP
.BR \-c ", "\-\-object
Output a relocatable object (old COFF format).
.TP
.BR \-C ", "\-\-new\-coff
Output a relocatable object (new COFF format).
.TP
.BR \-d ", " \-\-debug
Output debug messages.
.TP
.BR "\-D SYM=VAL" , " \-\-define SYM=VAL"
Define SYM with value VAL. This is equivalent to placing "#define SYM VAL" in
the source.
.TP
.BR "\-e [ON|OFF]" , " \-\-expand [ON|OFF]"
Macro expansion in the listing file is controlled using the EXPAND and NOEXPAND
directives.  This option can force either directive to be ignored.  When "ON"
is selected the macros are always expanded regardless of the presence of
NOEXPAND.  Likewise, "OFF" will force macros not to be expanded.
.TP
.BR \-g ", " \-\-debug-info
Use debug directives for COFF.
.TP
.BR \-h ", " \-\-help
Show the usage message.
.TP
.BR \-i ", " \-\-ignore\-case
All user defined symbols and macros are case sensitive.  This option makes them
case insensitive.
.TP
.BR "\-I DIR" , " \-\-include DIR"
The top level assembly file is passed to gpasm as an argument.  This file can
contain INCLUDE directives.  These directives open the specified file and
then its contents are assembled.  GPASM searches for these files in the local
directory.  If the file is not located, the default header path is checked.
Finally, all paths specified using this argument are searched.
.TP
.BR \-L ", "\-\-force-list
Ignore NOLIST directives.  This forces gpasm to print every line assembled in
the list file output.
.TP
.BR \-l ", "\-\-list\-chips
List the supported processors.
.TP
.BR \-m ", "\-\-dump
Upon completion of the final pass of assembly, display the contents of
instruction memory.
.TP
.BR \-M ", "\-\-deps
Output a dependency file.
.TP
.BR \-n ", "\-\-dos
By default, gpasm generates hex files using ISO format.  However, some device
programmers required a DOS formatted file.  This option will cause gpasm to
generate a DOS formatted hex file.
.TP
.BR "\-o FILE" , " \-\-output FILE"
Alternate name of the output file.
.TP
.BR "\-p PROC" , " \-\-processor PROC"
Select the processor.
.TP
.BR \-q ", "\-\-quiet
Suppress anything sent to standard output.
.TP
.BR "\-r RADIX" , " \-\-radix RADIX"
The supported radices are BIN, DEC, OCT, and HEX.  The default is HEX.
.TP
.BR \-v ", "\-\-version
Show the version.
.TP
.BR "\-w [0|1|2]" , " \-\-warning [0|1|2]"
This option sets the message level. "0" is the default.  It will allow all 
messages, warnings, and errors to be reported.  "1" will suppress the messages.
"2" will suppress the messages and warnings.
.TP
.BR \-y ", "\-\-extended
Enable 18xx extended mode.
.SH SEE ALSO
.BR gputils (1)
.SH AUTHOR
Craig Franklin <craigfranklin@users.sourceforge.net>
.SH COPYRIGHT
Copyright (C) 2002, 2003, 2004, 2005 James Bowman, Craig Franklin

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.


history.txt
======
snutil
======
Rev:		Date:		Reason:
0.0.3		12/6/29		eval version release	
0.1.0		12/6/29		sn16 class add 4 insn (inca, jza, deca, djza)
0.2.0		12/7/3		add cod output
0.3.0		12/7/6		add prefix S2-CONFIG BIT; S3-CHIP_ID in S19 output
0.4.0		12/7/10		sn16 class add 2 insn (rra, rla)
0.5.0		12/7/13		add coff output, add env param PATH 
0.6.0		12/8/22		correct "dt" assemble
0.6.1		12/10/23	add processor MC33P94, MC33P716, MC30P011
0.7.0		12/11/19	modify memonic rsubra, rsbcra etc.
0.8.0		12/12/5		sn16 class add 2 insn (lcall, lgoto); add processor MC32P21
0.8.1		12/12/14	correct lcall/lgoto opcode
0.9.0		13/1/20		correct lcall/lgoto link
1.0.0		13/3/19		add sn13 class; comra -> comr (sn14)
1.1.0		13/3/20		adcar -> adcra (sn13)
1.2.0		13/4/12		remove website
1.3.0		13/6/19		nessary update for c compiler
1.3.1		13/7/9		add processor MC32P64
1.4.0		13/9/16		correct relocate error; expand mc30 OPF7 insn address to 7bit
1.4.1_beta	13/10/15	add sn13b class, MC34P01
1.4.2_beta	14/1/6		sn13b opcode update, MC34P01
1.5.0		14/1/13		release sn13b class, MC34P01; remove adcai of sn16 class
1.6.0		14/1/22		add sn16b modify lcall/lgoto; remove lcall/lgoto of of other class; add processor MC32P821, MC33P78, MC33P116, MC30P081
1.6.1		14/1/23		correct gplink for lcall/lgoto
1.7.0_beta	14/4/11		add sn13c for MC31P5130
1.9.1
                加入新型号7323  之前MC32P7323用的是3221 在超空间1K报错。
        MC32P7323是8K的，超过8K报错，使用的是PRO_CLASS_16A内核，型号7323
1.9.0
                加入新内核 PRO_CLASS_SN16E      for     MC32T7333、MC9902
        MC32P7323是8K的，超过8K报错，使用的是PRO_CLASS_16A内核，型号7323
1.9.2	
		add processor 4K7323 for space 4K
	MC32P7323
1.9.3
		add processor 7212 for MC32P7212
1.9.4 	
		add processor 1K6220,05K6220 for MC30P6220	17/04/17
1.9.5
		add processor 1K7041,2K7041 for MC35P7041	17/04/18
		modify 7333 setup for MC32F7333			17/04/21
1.9.6
		add processor 7122	for MC32F7122		17/04/27
1.9.7		
		add processor 9903 9904 for A380		17/05/15	
1.9.8		
		add processor 7030 for MC32P7030		17/05/25
2.0.0_beta1 17/6/19	modify cod file for macro
2.0.0_beta2 17/6/23	output section location when link error ocurrs


gpasm/gpasm.c
/* top level functions for gpasm
   Copyright (C) 1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005
   James Bowman, Craig Franklin

This file is part of gputils.

gputils is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2, or (at your option)
any later version.

gputils is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with gputils; see the file COPYING.  If not, write to
the Free Software Foundation, 59 Temple Place - Suite 330,
Boston, MA 02111-1307, USA.  */

#include "stdhdr.h"

#include "libgputils.h"
#include "gpasm.h"
#include "gperror.h"
#include "scan.h"
#include "deps.h"
#include "directive.h"
#include "lst.h"
#include "cod.h"
#include "processor.h"
#include "coff.h"

struct gpasm_state state;

static gp_boolean cmd_processor = false;
static char *processor_name = NULL;

int yyparse(void);
extern int yydebug;

#define GET_OPTIONS "?D:I:a:cCde:ghilLmMno:p:qr:uvw:y"

static struct option longopts[] =
    {
        {"define", 1, 0, 'D'},
        {"include", 1, 0, 'I'},
        {"hex-format", 1, 0, 'a'},
        {"object", 0, 0, 'c'},
        {"new-coff", 0, 0, 'C'},
        {"debug", 0, 0, 'd'},
        {"expand", 1, 0, 'e'},
        {"debug-info", 0, 0, 'g'},
        {"help", 0, 0, 'h'},
        {"ignore-case", 0, 0, 'i'},
        {"list-chips", 0, 0, 'l'},
        {"force-list", 0, 0, 'L'},
        {"dump", 0, 0, 'm'},
        {"deps", 0, 0, 'M'},
        {"dos", 0, 0, 'n'},
        {"output", 1, 0, 'o'},
        {"processor", 1, 0, 'p'},
        {"quiet", 0, 0, 'q'},
        {"radix", 1, 0, 'r'},
        {"absolute", 0, 0, 'u'},
        {"version", 0, 0, 'v'},
        {"warning", 1, 0, 'w'},
        {"extended", 0, 0, 'y'},
        {0, 0, 0, 0}};

void init(void)
{

  gp_init();

  /* restore gpasm to its initialized state */
  state.mode = absolute;
  state.extended_pic16e = false;

  state.radix = 10;
  state.hex_format = inhx32;
  state.case_insensitive = false;
  state.quiet = false;
  state.use_absolute_path = false;
  state.error_level = 0;
  state.debug_info = false;
  state.path_num = 0;

  state.cmd_line.radix = false;
  state.cmd_line.hex_format = false;
  state.cmd_line.error_level = false;
  state.cmd_line.macro_expand = false;
  state.cmd_line.processor = false;
  state.cmd_line.lst_force = false;

  state.pass = 0;
  state.org = 0;
  state.device.id_location = 0;
  state.dos_newlines = false;
  state.memory_dump = false;
  state.found_config = false;
  state.found_devid = false;
  state.found_idlocs = false;
  state.found_end = false;
  state.maxram = (MAX_RAM - 1);

  state.codfile = normal;
  state.depfile = suppress;
  state.hexfile = normal;
  state.lstfile = normal;
  state.objfile = suppress;

  state.num.errors = 0;
  state.num.warnings = 0;
  state.num.messages = 0;
  state.num.warnings_suppressed = 0;
  state.num.messages_suppressed = 0;

  state.processor = no_processor;
  state.processor_chosen = 0;

  state.cod.enabled = false;
  state.dep.enabled = false;
  state.lst.enabled = false;
  state.obj.enabled = false;
  state.obj.newcoff = 1; /* use new Microchip COFF format by default */

  state.obj.object = NULL;
  state.obj.section = NULL;
  state.obj.symbol_num = 0;
  state.obj.section_num = 0;

  state.astack = NULL;

  state.next_state = state_nochange;

  return;
}

void add_path(const char *path)
{
  if (state.path_num < MAX_PATHS)
  {
    state.paths[state.path_num++] = strdup(path);
  }
  else
  {
    fprintf(stderr, "too many -I paths\n");
    exit(1);
  }
}

static void
show_usage(void)
{
  printf("Usage: gpasm [options] file\n");
  printf("Options: [defaults in brackets after descriptions]\n");
  /*
  printf("  -a FMT, --hex-format FMT       Select hex file format. [inhx32]\n");
  */
  printf("  -c, --object                   Output relocatable object.\n");
  /*
  printf("  -C, --old-coff                 Use old Microchip COFF format.\n");
  */
  printf("  -d, --debug                    Output debug messages.\n");
  printf("  -D SYM=VAL, --define SYM=VAL   Define SYM with value VAL.\n");
  printf("  -e [ON|OFF], --expand [ON|OFF] Macro expansion. [ON]\n");
  printf("  -g, --debug-info               Use debug directives for COFF.\n");
  printf("  -h, --help                     Show this usage message.\n");
  printf("  -i, --ignore-case              Case insensitive.\n");
  printf("  -I DIR, --include DIR          Specify include directory.\n");
  printf("  -l, --list-chips               List supported processors.\n");
  printf("  -L, --force-list               Ignore nolist directives.\n");
  printf("  -m, --dump                     Memory dump.\n");
  printf("  -M, --deps                     Output dependency file.\n");
#ifndef HAVE_DOS_BASED_FILE_SYSTEM
  printf("  -n, --dos                      Use DOS newlines in s19 file.\n");
#endif
  printf("  -o FILE, --output FILE         Alternate name of output file.\n");
  printf("  -p PROC, --processor PROC      Select processor.\n");
  printf("  -q, --quiet                    Quiet.\n");
  printf("  -r RADIX, --radix RADIX        Select radix. [dec]\n");
  /*  printf("  -u, --absolute                 Use absolute pathes. \n");*/
  printf("  -v, --version                  Show version.\n");
  printf("  -w [0|1|2], --warning [0|1|2]  Set message level. [0]\n");
  /*
  printf("  -y, --extended                 Enable 18xx extended mode.\n");
  */
  printf("\n");
#ifdef USE_DEFAULT_PATHS
  if (gp_header_path)
  {
    printf("Default header file path %s\n", gp_header_path);
  }
  else
  {
    printf("Default header file path NOT SET\n");
  }
  printf("\n");
#endif
  //printf("Report bugs to:\n");
  //printf("%s\n", PACKAGE_BUGREPORT);
}

void process_args(int argc, char *argv[])
{
  extern char *optarg;
  extern int optind;
  int c;
  gp_boolean usage = false;
  int usage_code = 0;
  char *pc;

  /* Scan through the options for the -i flag.  It must be set before the
     defines are read */
  while ((c = getopt_long(argc, argv, GET_OPTIONS, longopts, 0)) != EOF)
  {
    switch (c)
    {
    case 'i':
      state.case_insensitive = true;
      break;
    }
  }

  /* reset the getopt_long index for the next call */
  optind = 1;

  /* initalize the defines table for command line arguments */
  state.stDefines = push_symbol_table(NULL, state.case_insensitive);

  while ((c = getopt_long(argc, argv, GET_OPTIONS, longopts, 0)) != EOF)
  {
    switch (c)
    {
    case '?':
      usage_code = 1;
    case 'h':
      usage = true;
      break;
    /*
    case 'a':
      select_hexformat(optarg);
      state.cmd_line.hex_format = true;
      break;
	  */
    case 'c':
      state.mode = relocatable;
      state.codfile = suppress;
      state.hexfile = suppress;
      state.lstfile = normal;
      state.objfile = normal;
      break;
    /*
    case 'C':
      state.obj.newcoff = 0;
      break;
	  */
    case 'd':
      gp_debug_disable = false;
      break;
    case 'D':
      if ((optarg != NULL) && (strlen(optarg) > 0))
      {
        struct symbol *sym;
        char *lhs, *rhs;

        lhs = strdup(optarg);
        rhs = strchr(lhs, '=');
        if (rhs != NULL)
        {
          *rhs = '\0'; /* Terminate the left-hand side */
          rhs++;       /* right-hand side begins after the '=' */
        }

        sym = get_symbol(state.stDefines, lhs);
        if (sym == NULL)
          sym = add_symbol(state.stDefines, lhs);
        if (rhs)
          annotate_symbol(sym, rhs);
      }
      break;
    case 'e':
      select_expand(optarg);
      state.cmd_line.macro_expand = true;
      break;
    case 'g':
      state.debug_info = true;
      break;
    case 'I':
      add_path(optarg);
      break;
    case 'i':
      state.case_insensitive = true;
      break;
    case 'L':
      state.cmd_line.lst_force = true;
      break;
    case 'l':
      gp_dump_processor_list(true, 0);
      exit(0);
      break;
    case 'M':
      state.depfile = normal;
      break;
    case 'm':
      state.memory_dump = true;
      break;
    case 'n':
#ifndef HAVE_DOS_BASED_FILE_SYSTEM
      state.dos_newlines = true;
#endif
      break;
    case 'o':
      strncpy(state.objfilename, optarg, sizeof(state.objfilename));
      strncpy(state.basefilename, optarg, sizeof(state.basefilename));
      pc = strrchr(state.basefilename, '.');
      if (pc)
        *pc = 0;
      break;
    case 'p':
      cmd_processor = true;
      processor_name = optarg;
      break;
    case 'q':
      state.quiet = true;
      break;
    case 'r':
      select_radix(optarg);
      state.cmd_line.radix = true;
      break;
    /*    case 'u':
      state.use_absolute_path = true;
      break;*/
    case 'w':
      select_errorlevel(atoi(optarg));
      state.cmd_line.error_level = true;
      break;
    /*
    case 'y':
      state.extended_pic16e = true;
      break;
	  */
    case 'v':
      fprintf(stderr, "%s\n", GPASM_VERSION_STRING);
      exit(0);
    }
    if (usage)
      break;
  }

  if ((optind + 1) == argc)
    state.srcfilename = argv[optind];
  else
    usage = true;

  if (usage)
  {
    show_usage();
    exit(usage_code);
  }

  /* Add the header path to the include paths list last, so that the user
     specified directories are searched first */
  if (gp_header_path)
  {
    add_path(gp_header_path);
  }

  if (state.use_absolute_path)
  {
    state.srcfilename = gp_absolute_path(state.srcfilename);
  }
}

int assemble(void)
{
  char *pc;
  struct symbol_table *cmd_defines;

  /* store the command line defines to restore on second pass */
  cmd_defines = state.stDefines;
  state.c_memory = state.i_memory = i_memory_create();

  if (state.basefilename[0] == '\0')
  {
    strncpy(state.basefilename, state.srcfilename, sizeof(state.basefilename));
    pc = strrchr(state.basefilename, '.');
    if (pc)
      *pc = 0;
  }

  /* Builtins are always case insensitive */
  state.stBuiltin = push_symbol_table(NULL, true);
  state.stDirective = state.stBuiltin;
  state.stMacros = push_symbol_table(NULL, state.case_insensitive);
  state.stTop =
      state.stGlobal = push_symbol_table(NULL, state.case_insensitive);
  state.stTopDefines =
      state.stDefines = push_symbol_table(cmd_defines, state.case_insensitive);

  opcode_init(0);

  /* the tables are built, select the processor if -p was used */
  if (cmd_processor)
  {
    select_processor(processor_name);
    state.cmd_line.processor = true;
  }

  state.pass = 1;
  open_src(state.srcfilename, 0);
  yyparse();
  yylex_destroy();

  state.pass++;
  state.org = 0;
  state.device.id_location = 0;
  state.cblock = 0;
  state.cblock_defined = 0;
  /* clean out defines for second pass */
  state.stTopDefines =
      state.stDefines = push_symbol_table(cmd_defines, state.case_insensitive);
  purge_temp_symbols(state.stTop);
  if (!state.cmd_line.radix)
    state.radix = 10;
  state.obj.symbol_num = 0;
  state.obj.section_num = 0;

  /* Initial section for absolute mode */
  if (state.mode == absolute)
  {
    state.obj.new_sec_flags = STYP_TEXT;
  }

  state.found_config = false;
  state.found_devid = false;
  state.found_idlocs = false;
  state.found_end = false;
  coff_init();
  cod_init();
  deps_init();
  lst_init();

  /* Set F_EXTENDED18 COFF flag if 18xx extended mode enabled */
  if (state.obj.object && state.extended_pic16e)
    state.obj.object->flags |= F_EXTENDED18;

  open_src(state.srcfilename, 0);
  if (!gp_debug_disable)
  {
    yydebug = 1;
  }
  else
  {
    yydebug = 0;
  }
  yyparse();

  assert(state.pass == 2);

  pop_symbol_table(state.stBuiltin);

  printf("\n");
  hex_init();

  if (state.memory_dump)
    print_i_memory(state.i_memory, state.device.class);

  /* Maybe produce a symbol table */
  if (state.lst.symboltable)
  {
    lst_throw(); /* Start symbol table on a fresh page */
    lst_symbol_table(state.stGlobal);
    lst_defines_table(state.stDefines);
  }

  /* Maybe produce a memory map */
  if ((state.mode == absolute) && (state.lst.memorymap))
  {
    lst_memory_map(state.i_memory);
  }

  /* Finish off the object, dependency, listing, and symbol files*/
  coff_close_file();
  deps_close();
  lst_close();

  if (state.processor)
    cod_close_file();

  free_files();

  printf("\n");
  printf("%30s: %d\n", "errors", state.num.errors);
  printf("%30s: %d\n", "warnings", state.num.warnings);
  printf("%30s: %d\n", "messages", state.num.messages);
  printf("%30s: %d\n", "warnings suppressed", state.num.warnings_suppressed);
  printf("%30s: %d\n", "messages suppressed", state.num.messages_suppressed);
  printf("\n");

  if ((state.num.errors > 0) ||
      (gp_num_errors > 0))
  {
    printf("      ===== (GPASM) Unsuccessful =====\n\n");
    return EXIT_FAILURE;
  }
  else
  {
    printf("       ===== (GPASM) Successful =====\n\n");
    return EXIT_SUCCESS;
  }
}
